sklearn/metrics/_dist_metrics.pyx

```diff
@@ -554,27 +546,49 @@ cdef class MinkowskiDistance(DistanceMetric):
     r"""Minkowski Distance
 
     .. math::
-       D(x, y) = [\sum_i |x_i - y_i|^p] ^ (1/p)
+       D(x, y) = [\sum_i w_i *|x_i - y_i|^p] ^ (1/p)
+
+    Parameters
+    ----------
+    p : int
+        The order of the norm of the difference :math:`{||u-v||}_p`.
+    w : (N,) array-like (optional)
+        The weight vector.
 
     Minkowski Distance requires p >= 1 and finite. For p = infinity,
     use ChebyshevDistance.
     Note that for p=1, ManhattanDistance is more efficient, and for
     p=2, EuclideanDistance is more efficient.
     """
-    def __init__(self, p):
+    def __init__(self, p, w=None):
         if p < 1:
             raise ValueError("p must be greater than 1")
         elif np.isinf(p):
             raise ValueError("MinkowskiDistance requires finite p. "
                              "For p=inf, use ChebyshevDistance.")
+        elif w is not None and any(w_i < 0 for w_i in w):
+            raise ValueError("w cannot contain negative weights")
+
         self.p = p
+        self.vec = ReadonlyArrayWrapper(np.asarray([], dtype=DTYPE))
+        self.size = 0
+        if w is not None:
+            self.vec = ReadonlyArrayWrapper(np.asarray(w, dtype=DTYPE))
+            self.size = self.vec.shape[0]
+
+    def _validate_data(self, X):
+        if self.size > 0 and X.shape[1] != self.size:
+            raise ValueError('MinkowskiDistance dist: '
+                             'size of w does not match')
 
     cdef inline DTYPE_t rdist(self, const DTYPE_t* x1, const DTYPE_t* x2,
                               ITYPE_t size) nogil except -1:
-        cdef DTYPE_t d=0
+        cdef DTYPE_t vec_j, d=0
         cdef np.intp_t j
+        cdef bint has_w = self.size > 0
         for j in range(size):
-            d += pow(fabs(x1[j] - x2[j]), self.p)
+            vec_j = self.vec[j] if has_w else 1.
+            d += vec_j * pow(fabs(x1[j] - x2[j]), self.p)
```

Comment:
To avoid the speed penalty if testing for has_w at each iteration, please move the condition around the loop to have a specialized loop for the common weightless case:

```python
        if has_w:
            for j in range(size):
                d += self.vec[j] * pow(fabs(x1[j] - x2[j]), self.p)
        else:
            for j in range(size):
                d += pow(fabs(x1[j] - x2[j]), self.p)
```